### PR TITLE
Add the proj_positive

### PR DESCRIPTION
**Pull request by nperraud**
This is an example for Vassilis

### CODE REVIEW

**Comment by nperraud** [comment]
The test is not passing, but I do not think it is my fault.

**Comment by ThomasA** [comment]
The build seems to be failing, because someone is trying to instantiate a class in `pyunlocbox.functions` that is not implemented. If someone could figure out why the tests are failing that would be really helpful to start working on #11.

**Comment by mdeff** [comment]
@nperraud I'm sorry to inform you it's your fault :P

The issue is that the `test_independent_problems` function calls the gradient of every function, but yours don't define one (rightly, it doesn't have one). You shall thus add it to the list of functions to be ignored [there](https://github.com/epfl-lts2/pyunlocbox/blob/6b8f75b6d7a7aa6db33b138cea56ea3456c5abcf/pyunlocbox/tests/test_functions.py#L417). A better way is probably to modify that function so that it calls the `cap` method to know the capabilities (`'EVAL', 'GRAD', 'PROX'`) of a particular function object.

Review:
* write in `HISTORY.md` that you added this prox
* fix the style issues. From the Travis output:
```
$ make lint
flake8 --doctests --exclude=doc
./pyunlocbox/functions.py:824:1: E302 expected 2 blank lines, found 1
./pyunlocbox/functions.py:859:1: E302 expected 2 blank lines, found 1
./pyunlocbox/tests/test_functions.py:379:5: E303 too many blank lines (2)
./pyunlocbox/tests/test_functions.py:386:9: E303 too many blank lines (2)
./pyunlocbox/tests/test_functions.py:387:31: E231 missing whitespace after ','
./pyunlocbox/tests/test_functions.py:389:26: E231 missing whitespace after ','
./pyunlocbox/tests/test_functions.py:391:28: E225 missing whitespace around operator
./pyunlocbox/tests/test_functions.py:394:30: E225 missing whitespace around operator
./pyunlocbox/tests/test_functions.py:394:33: E225 missing whitespace around operator
./pyunlocbox/tests/test_functions.py:397:34: E225 missing whitespace around operator
./pyunlocbox/tests/test_functions.py:397:37: E231 missing whitespace after ','
./pyunlocbox/tests/test_functions.py:397:41: E225 missing whitespace around operator
./pyunlocbox/tests/test_functions.py:403:5: E303 too many blank lines (2)
make: *** [lint] Error 1
```

PS: you can run the tests locally with `make test`, `make lint` and `make doc`.

**Comment by nperraud** [comment]
Thanks for this very clear an precise answer.

**Comment by nperraud** [comment]
I have change the way the test of the function is done. Now, it is supposed to test if the function has a prox, a grad and and eval function. It is not perfect but it should be better. 

The coverage seems to be a bit reduced. So maybe, I forgot a case.

I have also change the test for the norm_tv. Now it works a bit better.

File: pyunlocbox/tests/test_functions.py [hunk 1]
```diff
@@ -413,37 +410,59 @@ def test_independent_problems(self):
         # Test all available functions.
         funcs = inspect.getmembers(functions, inspect.isclass)
         for func in funcs:
+            print('Testing ' + func[0])
```
Comment (mdeff) [inline pyunlocbox/tests/test_functions.py]: Please don't leave `print` statements in tests. It's useful for debugging but it makes the test log noisy. You can know where in the loop the test failed by wrapping the code in a [`with self.subTest(...):`](https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests) block.

File: pyunlocbox/tests/test_functions.py [hunk 2]
```diff
@@ -413,37 +410,59 @@ def test_independent_problems(self):
...
                 res = 0
                 for iN in range(N):
                     res += f.eval(X[:, iN])
                 nptest.assert_array_almost_equal(res, f.eval(X))
 
             # Each column is the prox of one of the N problems.
             # TODO: norm_tv shoud pass this test. Why is there a difference ?
-            if func[0] not in ['func', 'norm', 'norm_nuclear', 'norm_tv',
-                               'proj']:
+            if has_prox:
```
Comment (mdeff) [inline pyunlocbox/tests/test_functions.py]: There's a method for this already. You can use `if 'PROX' in func[0].cap()`. Please do the same for `has_eval` and `has_grad` and remove those 3 variables.

File: pyunlocbox/tests/test_functions.py [hunk 3]
```diff
@@ -413,37 +410,59 @@ def test_independent_problems(self):
...
+                assert (has_eval)
+                # Assert if the function has a gradient or a proximal op.
+                assert (has_grad or has_prox)
+
 
 suite = unittest.TestLoader().loadTestsFromTestCase(TestCase)
+
+
+if __name__ == '__main__':
+    unittest.main()
```
Comment (mdeff) [inline pyunlocbox/tests/test_functions.py]: No need for this. We can run part of a test suite with e.g. `python -m unittest pyunlocbox.tests.test_functions`.

File: pyunlocbox/tests/test_functions.py [hunk 4]
```diff
@@ -413,37 +410,59 @@ def test_independent_problems(self):
...
-            if func[0] not in ['func', 'norm', 'norm_l1', 'norm_nuclear',
-                               'norm_tv', 'proj', 'proj_b2']:
+            if has_grad:
                 res = np.zeros((n, N))
                 for iN in range(N):
                     res[:, iN] = f.grad(X[:, iN])
                 nptest.assert_array_almost_equal(res, f.grad(X))
 
+            # TODO: this test should probably be done in a more clever way
+            if func[0] not in ['norm', 'proj', 'proj_b2', 'proj_positive'
```
Comment (mdeff) [inline pyunlocbox/tests/test_functions.py]: Why is this test useful?

**Comment by nperraud** [comment]
I updated the test cases according to your recommendation. I am just not so sure about the cap function. Since it has a try, except statement, are we going to skip some tests that are just failing?

I commented my last test to explain what I had in mind.

**Comment by nperraud** [comment]
It seems that self.subtest is not working with python 2. Shall I just remove it?

**Comment by mdeff** [comment]
Thanks @nperraud :+1:

Never mind about `self.subTest`, we can add it later when we'll drop support for python 2.7.

> I am just not so sure about the cap function. Since it has a try, except statement, are we going to skip some tests that are just failing?

Yep but that's not the purpose of the `test_independent_problems` test. This function simply tests that multiple independent problems can be solved in parallel. What a particular function implements (and if it does it properly) is tested in their specific test function, like the `test_proj_positive` you implemented.

> Test if the function makes sense, i.e: the functin should contain the function _eval and one of the function _prox or _grad.

Likewise, this should not be in the `test_independent_problems` test, but rather in another function, as it performs another test.

**Comment by nperraud** [comment]
The coveralls decrease comes from the fact that I fixed the test for the tv projection... I think it is time to merge.

**Comment by mdeff** [comment]
No prob for decreasing coverage.

Can we just put the test that checks if a function makes sense, i.e. the function should contain the function _eval and one of the function _prox or _grad, in another function than `test_independent_problems`? Thanks

**Comment by nperraud** [comment]
@mdeff do you agree that I merge this?
It would solve issue #20

**Comment by mdeff** [comment]
I fixed the remaining issues. Thanks for the PR!

**Pull request merged**